src/zenml/new/pipelines/pipeline.py

```diff
@@ -0,0 +1,1088 @@
+            input_artifacts: The input artifacts for the invocation.
+            external_artifacts: The external artifacts for the invocation.
+            parameters: The parameters for the invocation.
+            upstream_steps: The upstream steps for the invocation.
+            custom_id: Custom ID to use for the invocation.
+            allow_id_suffix: Whether a suffix can be appended to the invocation
+                ID.
+
+        Raises:
+            RuntimeError: If the method is called on an inactive pipeline.
+            RuntimeError: If the invocation was called with an artifact from
+                a different pipeline.
+
+        Returns:
+            The step invocation ID.
+        """
+        if Pipeline.ACTIVE_PIPELINE != self:
+            raise RuntimeError(
+                "A step invocation can only be added to an active pipeline."
+            )
+
+        for artifact in input_artifacts.values():
+            if artifact.pipeline is not self:
+                raise RuntimeError(
+                    "Got invalid input artifact for invocation of step "
+                    f"{step.name}: The input artifact was produced by a step "
+                    f"inside a different pipeline {artifact.pipeline.name}."
+                )
+
+        invocation_id = self._compute_invocation_id(
+            step=step, custom_id=custom_id, allow_suffix=allow_id_suffix
+        )
+        invocation = StepInvocation(
+            id=invocation_id,
+            step=step,
+            input_artifacts=input_artifacts,
+            external_artifacts=external_artifacts,
+            parameters=parameters,
+            upstream_steps=upstream_steps,
+            pipeline=self,
+        )
+        self._invocations[invocation_id] = invocation
+        return invocation_id
+
+    def _compute_invocation_id(
+        self,
+        step: "BaseStep",
+        custom_id: Optional[str] = None,
+        allow_suffix: bool = True,
+    ) -> str:
+        """Compute the invocation ID.
+
+        Args:
+            step: The step for which to compute the ID.
+            custom_id: Custom ID to use for the invocation.
+            allow_suffix: Whether a suffix can be appended to the invocation
+                ID.
+
+        Raises:
+            RuntimeError: If no ID suffix is allowed and an invocation for the
+                same ID already exists.
+            RuntimeError: If no unique invocation ID can be found.
+
+        Returns:
+            The invocation ID.
+        """
+        base_id = custom_id or step.name
+
+        if base_id in self.invocations and not allow_suffix:
+            raise RuntimeError("Duplicate step ID")
+
+        id_ = base_id
+        for index in range(2, 10000):
+            if id_ not in self.invocations:
+                break
+
+            id_ = f"{base_id}_{index}"
+        else:
+            raise RuntimeError("Unable to find step ID")
+
+        return id_
```

Comment:
A bit more readable:

```suggestion
        base_id = id_ = custom_id or step.name

        if id_ not in self.invocations:
            return id_

        if not allow_suffix:
            raise RuntimeError("Duplicate step ID")

        for index in range(2, 10000):
            id_ = f"{base_id}_{index}"
            if id_ not in self.invocations:
                return id_

        raise RuntimeError("Unable to find step ID")
```